### DISCUSSION TITLE
What is the best way to get all the messages in a channel? My current method is consistently getting me rate limited every other message.

### DISCUSSION THREAD

**Discussion by quinton-oglesby** [shardlab/discordrb#129] [Q&A] [answered]
<img width="495" alt="Screen Shot 2022-01-30 at 11 52 03 AM" src="https://user-images.githubusercontent.com/70076973/151711137-eebb4195-6f66-46be-bb65-881764aa4891.png">
My current method of doing so looks like this, but surely there is a better way of doing this? I have 1.5 million messages I'm trying to scrape and if I keep having to put my bot to sleep every five seconds to avoid rate limiting then that is going to take nearly 90 days.

**Comment by swarley** [comment, marked as answer]
Answered in discord
```ruby
bot = Discordrb::Bot.new(token: ENV['INOSUKE_TOKEN']);
after_id = 0
message_store = []

while (messages = bot.channel(345687437722386433).history(100, nil, after_id)).count == 100
  after_id = messages.map(&:id).max  
  puts "new after #{after_id}"  
  message_store += messages # or do whatever you want to do with the messages to avoid keeping them in memory 
end

# and you'll need to handle the last set of messages again out here, or use a flag set inside of the loop for continuing instead
```
